src/ZkFold/Cardano/OffChain/Transcript.hs

```diff
@@ -0,0 +1,37 @@
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module ZkFold.Cardano.OffChain.Transcript where
+
+import           Data.Word                                   (Word8)
+import           GHC.ByteOrder                               (ByteOrder (..))
+import           PlutusTx.Builtins
+import           PlutusTx.Prelude                            ((.))
+
+import           ZkFold.Base.Algebra.Basic.Field             (toZp)
+import           ZkFold.Base.Algebra.EllipticCurve.BLS12_381 (BLS12_381_G1, Fr)
+import           ZkFold.Base.Algebra.EllipticCurve.Class     (PointCompressed)
+import           ZkFold.Base.Data.ByteString                 (toByteString)
+import           ZkFold.Base.Protocol.NonInteractiveProof    (FromTranscript (..), ToTranscript (..))
+import           ZkFold.Cardano.OffChain.ECC                 (convertZp)
+import           ZkFold.Cardano.OnChain.BLS12_381
+
+instance ToTranscript BuiltinByteString Word8 where
+    toTranscript = toBuiltin . toByteString
+
+instance ToTranscript BuiltinByteString F where
+    toTranscript (F n) = integerToByteString LittleEndian 32 n
+
+instance ToTranscript BuiltinByteString G1 where
+    toTranscript = bls12_381_G1_compress
+
+instance FromTranscript BuiltinByteString F where
+    fromTranscript = F . byteStringToInteger LittleEndian . blake2b_224
+
+instance ToTranscript BuiltinByteString Fr where
+    toTranscript = integerToByteString LittleEndian 32 . convertZp
```

Comment:
I can't understand why we have F and Fr. Isn't it the same thing?

Reply:
No, they are not. `F` is defined in `zkfold-cardano` and is based on `Integer` while `Fr` is based on `Natural` and is defined in `zkfold-base`.

Reply:
So, we can't use `Fr` in our on-chain code directly. Plus, in the future `Fr` might be based on `ByteString`.

Reply:
Thank you, now I see how it happened here